File: tom_som_javascript_runtime/tom_som_runtime/spec_meta_diff.js
'use strict';

/**
 * Structural comparison of two {@link SomMetaNode} subtrees; a faithful port
 * of `tom_som_dart_runtime/lib/src/spec_meta_diff.dart` (and the Python/Java
 * ports).
 *
 * {@link somMetaNodeDiff} is the agreement oracle for DR8: the generated
 * facades embed populated metadata trees as static code (DR1 §3.2), while
 * {@link module:spec_meta_bridge.buildSomMetaTree} derives the same tree from
 * the exported meta-JSON at runtime — the two must be field-for-field
 * identical for every node. Tests compare them with this function, which
 * returns a human-readable description of the **first** difference found
 * (with the node's position), or `null` when the subtrees agree completely.
 */

/**
 * Compares the `a` and `b` subtrees field by field (annotations, names,
 * kinds, form/document metadata, children and list element subtrees).
 *
 * Returns `null` when the subtrees are structurally identical, else a
 * description of the first difference, prefixed with the position `at`
 * (a `/`-joined member-name chain, defaulting to the root marker).
 *
 * @param {import('./spec_meta').SomMetaNode} a
 * @param {import('./spec_meta').SomMetaNode} b
 * @param {string} [at]
 * @returns {string|null}
 */
function somMetaNodeDiff(a, b, at = '<root>') {
  function diff(field, va, vb) {
    return _valueEq(va, vb) ? null : `${at}: ${field} differs — ${va} != ${vb}`;
  }

  const checks = [
    () => diff('className', a.className, b.className),
    () => diff('memberName', a.memberName, b.memberName),
    () => diff('sectionId', a.sectionId, b.sectionId),
    () => diff('sectionIdPattern', a.sectionIdPattern, b.sectionIdPattern),
    () => diff('kind', a.kind, b.kind),
    () => diff('typeName', a.typeName, b.typeName),
    () =>
      diff('serializationOrder', a.serializationOrder, b.serializationOrder),
    () => diff('min', a.min, b.min),
    () => diff('unused', a.unused, b.unused),
    () =>
      diff(
        'contentType.type',
        a.contentType ? a.contentType.type : null,
        b.contentType ? b.contentType.type : null,
      ),
    () =>
      diff(
        'contentType.description',
        a.contentType ? a.contentType.description : null,
        b.contentType ? b.contentType.description : null,
      ),
    () => diff('contentHelp', a.contentHelp, b.contentHelp),
    () => diff('headline', a.headline, b.headline),
    () => diff('comment', a.comment, b.comment),
    () => diff('docComment', a.docComment, b.docComment),
    () => diff('classDocComment', a.classDocComment, b.classDocComment),
    () => diff('mapsTo', a.mapsTo, b.mapsTo),
    () => diff('detailedIn', a.detailedIn, b.detailedIn),
    () => diff('recursive', a.recursive, b.recursive),
    () => _formDiff(at, a.form, b.form),
    () => _documentDiff(at, a.document, b.document),
    () => _secondLevelDiff(at, a.secondLevelIds, b.secondLevelIds),
    () => _extraDiff(at, a.extra, b.extra),
  ];
  for (const check of checks) {
    const d = check();
    if (d !== null) {
      return d;
    }
  }

  if (a.children.length !== b.children.length) {
    return (
      `${at}: children count differs — ` +
      `${a.children.length} != ${b.children.length} ` +
      `(${JSON.stringify(a.children.map((c) => c.memberName))} vs ` +
      `${JSON.stringify(b.children.map((c) => c.memberName))})`
    );
  }
  for (let i = 0; i < a.children.length; i++) {
    const ca = a.children[i];
    const d = somMetaNodeDiff(
      ca,
      b.children[i],
      `${at}/${ca.memberName !== null ? ca.memberName : ca.className}`,
    );
    if (d !== null) {
      return d;
    }
  }

  const aElem = a.elementNode !== null && a.elementNode !== undefined;
  const bElem = b.elementNode !== null && b.elementNode !== undefined;
  if (aElem !== bElem) {
    return `${at}: elementNode presence differs — ${aElem} != ${bElem}`;
  }
  if (aElem) {
    return somMetaNodeDiff(a.elementNode, b.elementNode, `${at}/§element`);
  }
  return null;
}

function _formDiff(at, a, b) {
  const aSet = a !== null && a !== undefined;
  const bSet = b !== null && b !== undefined;
  if (aSet !== bSet) {
    return `${at}: form presence differs — ${aSet} != ${bSet}`;
  }
  if (!aSet) {
    return null;
  }
  if (a.fields.length !== b.fields.length) {
    return (
      `${at}: form field count differs — ` +
      `${a.fields.length} != ${b.fields.length}`
    );
  }
  for (let i = 0; i < a.fields.length; i++) {
    const fa = a.fields[i];
    const fb = b.fields[i];
    if (
      fa.name !== fb.name ||
      fa.typeName !== fb.typeName ||
      fa.description !== fb.description ||
      fa.required !== fb.required ||
      fa.hint !== fb.hint ||
      fa.role !== fb.role ||
      fa.initial !== fb.initial ||
      fa.order !== fb.order
    ) {
      return `${at}: form field ${fa.name} differs`;
    }
  }
  return null;
}

function _documentDiff(at, a, b) {
  const aSet = a !== null && a !== undefined;
  const bSet = b !== null && b !== undefined;
  if (aSet !== bSet) {
    return `${at}: document presence differs — ${aSet} != ${bSet}`;
  }
  if (!aSet) {
    return null;
  }
  if (a.name !== b.name) {
    return `${at}: document.name differs — ${a.name} != ${b.name}`;
  }
  if (a.description !== b.description) {
    return `${at}: document.description differs`;
  }
  if (!_listEq(a.basedOn, b.basedOn)) {
    return (
      `${at}: document.basedOn differs — ` +
      `${JSON.stringify(a.basedOn)} != ${JSON.stringify(b.basedOn)}`
    );
  }
  return null;
}

function _secondLevelDiff(at, a, b) {
  if (a.length !== b.length) {
    return `${at}: secondLevelIds count differs — ${a.length} != ${b.length}`;
  }
  for (let i = 0; i < a.length; i++) {
    if (a[i].documentClass !== b[i].documentClass || a[i].id !== b[i].id) {
      return `${at}: secondLevelIds[${i}] differs`;
    }
  }
  return null;
}

function _extraDiff(at, a, b) {
  if (a.length !== b.length) {
    return (
      `${at}: extra annotation count differs — ${a.length} != ${b.length} ` +
      `(${JSON.stringify(a.map((e) => e.annotation))} vs ` +
      `${JSON.stringify(b.map((e) => e.annotation))})`
    );
  }
  for (let i = 0; i < a.length; i++) {
    if (
      a[i].annotation !== b[i].annotation ||
      !_valueEq(a[i].args, b[i].args)
    ) {
      return (
        `${at}: extra annotation ${a[i].annotation} differs — ` +
        `${JSON.stringify(a[i].args)} != ${JSON.stringify(b[i].args)}`
      );
    }
  }
  return null;
}

/** Deep structural equality over JSON-shaped values (the annotation-argument
 *  shapes: scalars, arrays, plain objects). */
function _valueEq(a, b) {
  if (Array.isArray(a) && Array.isArray(b)) {
    return _listEq(a, b);
  }
  if (_isPlainObject(a) && _isPlainObject(b)) {
    const aKeys = Object.keys(a);
    if (aKeys.length !== Object.keys(b).length) {
      return false;
    }
    for (const k of aKeys) {
      if (!Object.prototype.hasOwnProperty.call(b, k) || !_valueEq(a[k], b[k])) {
        return false;
      }
    }
    return true;
  }
  return a === b;
}

function _listEq(a, b) {
  if (a.length !== b.length) {
    return false;
  }
  for (let i = 0; i < a.length; i++) {
    if (!_valueEq(a[i], b[i])) {
      return false;
    }
  }
  return true;
}

function _isPlainObject(v) {
  return v !== null && typeof v === 'object' && !Array.isArray(v);
}

module.exports = {
  somMetaNodeDiff,
};

File: tom_som_rust_runtime/src/spec_meta_diff.rs
//! `spec_meta_diff` — structural comparison of two [`SomMetaNode`] subtrees; a
//! faithful port of `tom_som_go_runtime/spec_meta_diff.go` (itself a port of
//! `tom_som_dart_runtime/lib/src/spec_meta_diff.dart` and the TypeScript /
//! Python / JavaScript siblings).
//!
//! [`som_meta_node_diff`] is the agreement oracle: the generated
//! facades embed populated metadata trees as static code (SOM §7.2), while
//! [`crate::build_som_meta_tree`] derives the same tree from the exported
//! meta-JSON at runtime — the two must be field-for-field identical for every
//! node. Tests compare them with this function, which returns a human-readable
//! description of the **first** difference found (with the node's position),
//! or the empty string when the subtrees agree completely.
//!
//! Rust conventions (documented divergences from the Go port): optional ints
//! are `Option<i64>` instead of `*int` (still rendered as `null` when absent so
//! the message strings stay aligned with the sibling ports), and
//! annotation-argument values are the runtime's own [`Json`] enum, whose
//! numeric variants compare by numeric value across `Int`/`Float` (the JSON
//! decoder yields `Float` where generated literals may be integral).

use crate::json::{encode_str, Json};
use crate::spec_meta::{
    SomContentTypeMeta, SomDocMeta, SomFormMeta, SomMetaExtra, SomMetaNode,
};

/// Compares the `a` and `b` subtrees field by field (annotations, names,
/// kinds, form/document metadata, children and list element subtrees).
///
/// Returns `""` when the subtrees are structurally identical, else a
/// description of the first difference, prefixed with the node's position (a
/// `/`-joined member-name chain rooted at the `<root>` marker).
pub fn som_meta_node_diff(a: &SomMetaNode, b: &SomMetaNode) -> String {
    som_meta_node_diff_at(a, b, "<root>")
}

fn som_meta_node_diff_at(a: &SomMetaNode, b: &SomMetaNode, at: &str) -> String {
    if a.class_name != b.class_name {
        return str_diff(at, "className", &a.class_name, &b.class_name);
    }
    if a.member_name != b.member_name {
        return str_diff(at, "memberName", &a.member_name, &b.member_name);
    }
    if a.section_id != b.section_id {
        return str_diff(at, "sectionId", &a.section_id, &b.section_id);
    }
    if a.section_id_pattern != b.section_id_pattern {
        return str_diff(
            at,
            "sectionIdPattern",
            &a.section_id_pattern,
            &b.section_id_pattern,
        );
    }
    if a.kind != b.kind {
        return str_diff(at, "kind", &a.kind, &b.kind);
    }
    if a.type_name != b.type_name {
        return str_diff(at, "typeName", &a.type_name, &b.type_name);
    }
    if a.serialization_order != b.serialization_order {
        return str_diff(
            at,
            "serializationOrder",
            &opt_int_repr(a.serialization_order),
            &opt_int_repr(b.serialization_order),
        );
    }
    if a.min != b.min {
        return str_diff(at, "min", &opt_int_repr(a.min), &opt_int_repr(b.min));
    }
    if a.unused != b.unused {
        return str_diff(
            at,
            "unused",
            &a.unused.to_string(),
            &b.unused.to_string(),
        );
    }
    let (a_ct_type, a_ct_desc) = content_type_fields(&a.content_type);
    let (b_ct_type, b_ct_desc) = content_type_fields(&b.content_type);
    if a_ct_type != b_ct_type {
        return str_diff(at, "contentType.type", &a_ct_type, &b_ct_type);
    }
    if a_ct_desc != b_ct_desc {
        return str_diff(at, "contentType.description", &a_ct_desc, &b_ct_desc);
    }
    if a.content_help != b.content_help {
        return str_diff(at, "contentHelp", &a.content_help, &b.content_help);
    }
    if a.headline != b.headline {
        return str_diff(at, "headline", &a.headline, &b.headline);
    }
    if a.comment != b.comment {
        return str_diff(at, "comment", &a.comment, &b.comment);
    }
    if a.doc_comment != b.doc_comment {
        return str_diff(at, "docComment", &a.doc_comment, &b.doc_comment);
    }
    if a.class_doc_comment != b.class_doc_comment {
        return str_diff(
            at,
            "classDocComment",
            &a.class_doc_comment,
            &b.class_doc_comment,
        );
    }
    if a.maps_to != b.maps_to {
        return str_diff(at, "mapsTo", &a.maps_to, &b.maps_to);
    }
    if a.detailed_in != b.detailed_in {
        return str_diff(at, "detailedIn", &a.detailed_in, &b.detailed_in);
    }
    if a.recursive != b.recursive {
        return str_diff(
            at,
            "recursive",
            &a.recursive.to_string(),
            &b.recursive.to_string(),
        );
    }
    let d = meta_form_diff(at, &a.form, &b.form);
    if !d.is_empty() {
        return d;
    }
    let d = meta_document_diff(at, &a.document, &b.document);
    if !d.is_empty() {
        return d;
    }
    let d = meta_extra_diff(at, &a.extra, &b.extra);
    if !d.is_empty() {
        return d;
    }

    if a.children.len() != b.children.len() {
        return format!(
            "{}: children count differs — {} != {} ({} vs {})",
            at,
            a.children.len(),
            b.children.len(),
            json_repr_strings(&member_names(&a.children)),
            json_repr_strings(&member_names(&b.children)),
        );
    }
    for (ca, cb) in a.children.iter().zip(b.children.iter()) {
        let name = if ca.member_name.is_empty() {
            &ca.class_name
        } else {
            &ca.member_name
        };
        let d = som_meta_node_diff_at(ca, cb, &format!("{}/{}", at, name));
        if !d.is_empty() {
            return d;
        }
    }

    let a_elem = a.element_node.is_some();
    let b_elem = b.element_node.is_some();
    if a_elem != b_elem {
        return format!(
            "{}: elementNode presence differs — {} != {}",
            at, a_elem, b_elem
        );
    }
    if let (Some(ea), Some(eb)) = (&a.element_node, &b.element_node) {
        return som_meta_node_diff_at(ea, eb, &format!("{}/§element", at));
    }
    String::new()
}

fn str_diff(at: &str, field: &str, a: &str, b: &str) -> String {
    format!("{}: {} differs — {} != {}", at, field, a, b)
}

fn member_names(nodes: &[std::rc::Rc<SomMetaNode>]) -> Vec<String> {
    nodes.iter().map(|n| n.member_name.clone()).collect()
}

/// Renders the two comparable fields of an optional `@ContentType` the way the
/// sibling ports interpolate them (`null` for an absent annotation).
fn content_type_fields(ct: &Option<SomContentTypeMeta>) -> (String, String) {
    match ct {
        None => ("null".to_string(), "null".to_string()),
        Some(c) => (c.type_.clone(), c.description.clone()),
    }
}

fn opt_int_repr(v: Option<i64>) -> String {
    match v {
        None => "null".to_string(),
        Some(n) => n.to_string(),
    }
}

fn meta_form_diff(at: &str, a: &Option<SomFormMeta>, b: &Option<SomFormMeta>) -> String {
    let a_set = a.is_some();
    let b_set = b.is_some();
    if a_set != b_set {
        return format!("{}: form presence differs — {} != {}", at, a_set, b_set);
    }
    let (Some(fa), Some(fb)) = (a, b) else {
        return String::new();
    };
    if fa.fields.len() != fb.fields.len() {
        return format!(
            "{}: form field count differs — {} != {}",
            at,
            fa.fields.len(),
            fb.fields.len()
        );
    }
    for (x, y) in fa.fields.iter().zip(fb.fields.iter()) {
        if x.name != y.name
            || x.type_name != y.type_name
            || x.description != y.description
            || x.required != y.required
            || x.hint != y.hint
            || x.order != y.order
        {
            return format!("{}: form field {} differs", at, x.name);
        }
    }
    String::new()
}

fn meta_document_diff(at: &str, a: &Option<SomDocMeta>, b: &Option<SomDocMeta>) -> String {
    let a_set = a.is_some();
    let b_set = b.is_some();
    if a_set != b_set {
        return format!(
            "{}: document presence differs — {} != {}",
            at, a_set, b_set
        );
    }
    let (Some(da), Some(db)) = (a, b) else {
        return String::new();
    };
    if da.name != db.name {
        return format!("{}: document.name differs — {} != {}", at, da.name, db.name);
    }
    if da.description != db.description {
        return format!("{}: document.description differs", at);
    }
    if da.based_on != db.based_on {
        return format!(
            "{}: document.basedOn differs — {} != {}",
            at,
            json_repr_strings(&da.based_on),
            json_repr_strings(&db.based_on)
        );
    }
    String::new()
}

fn meta_extra_diff(at: &str, a: &[SomMetaExtra], b: &[SomMetaExtra]) -> String {
    if a.len() != b.len() {
        let a_names: Vec<String> = a.iter().map(|e| e.annotation.clone()).collect();
        let b_names: Vec<String> = b.iter().map(|e| e.annotation.clone()).collect();
        return format!(
            "{}: extra annotation count differs — {} != {} ({} vs {})",
            at,
            a.len(),
            b.len(),
            json_repr_strings(&a_names),
            json_repr_strings(&b_names)
        );
    }
    for (x, y) in a.iter().zip(b.iter()) {
        if x.annotation != y.annotation || !args_eq(&x.args, &y.args) {
            return format!(
                "{}: extra annotation {} differs — {} != {}",
                at,
                x.annotation,
                json_repr_args(&x.args),
                json_repr_args(&y.args)
            );
        }
    }
    String::new()
}

/// Compares two annotation-argument lists as unordered string-keyed maps —
/// mirroring the Go port, whose args are `map[string]interface{}` (a nil map
/// equals an empty one; here both are just empty `Vec`s).
fn args_eq(a: &[(String, Json)], b: &[(String, Json)]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    a.iter().all(|(k, av)| {
        b.iter()
            .find(|(bk, _)| bk == k)
            .is_some_and(|(_, bv)| meta_value_eq(av, bv))
    })
}

/// Deep structural equality over JSON-shaped values (the annotation-argument
/// shapes: scalars, arrays, string-keyed objects). Numbers compare by numeric
/// value across `Int`/`Float` (see the module header); objects compare as
/// unordered maps, mirroring Go's `map[string]interface{}` semantics.
fn meta_value_eq(a: &Json, b: &Json) -> bool {
    if let (Some(an), Some(bn)) = (as_float(a), as_float(b)) {
        return an == bn;
    }
    match (a, b) {
        (Json::Array(al), Json::Array(bl)) => {
            al.len() == bl.len()
                && al.iter().zip(bl.iter()).all(|(x, y)| meta_value_eq(x, y))
        }
        (Json::Object(am), Json::Object(bm)) => args_eq(am, bm),
        _ => a == b,
    }
}

fn as_float(v: &Json) -> Option<f64> {
    match v {
        Json::Int(n) => Some(*n as f64),
        Json::Float(n) => Some(*n),
        _ => None,
    }
}

/// Renders a string list as compact JSON for diff messages (mirroring the Go
/// port's `json.Marshal` interpolations).
fn json_repr_strings(items: &[String]) -> String {
    let parts: Vec<String> = items.iter().map(|s| encode_str(s)).collect();
    format!("[{}]", parts.join(","))
}

/// Renders an annotation-argument list as a compact JSON object.
fn json_repr_args(args: &[(String, Json)]) -> String {
    let parts: Vec<String> = args
        .iter()
        .map(|(k, v)| format!("{}:{}", encode_str(k), json_repr(v)))
        .collect();
    format!("{{{}}}", parts.join(","))
}

/// Renders a [`Json`] value as compact JSON for diff messages.
fn json_repr(v: &Json) -> String {
    match v {
        Json::Null => "null".to_string(),
        Json::Bool(b) => b.to_string(),
        Json::Int(n) => n.to_string(),
        Json::Float(n) => format!("{}", n),
        Json::Str(s) => encode_str(s),
        Json::Array(items) => {
            let parts: Vec<String> = items.iter().map(json_repr).collect();
            format!("[{}]", parts.join(","))
        }
        Json::Object(pairs) => json_repr_args(pairs),
    }
}
